healthchecker/grpc.go

```diff
@@ -46,3 +50,23 @@ func (hc *grpcHealthChecker) Fail() {
 func (hc *grpcHealthChecker) GetServer() *health.Server {
 	return hc.srv
 }
+
+func GrpcUnaryServerHealthCheckerInterceptor(sigs chan os.Signal, failedHealthCheck chan struct{}) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		logger := contextutils.LoggerFrom(ctx)
+
+		select {
+		case x, ok := <-sigs:
+			if ok {
+				logger.Debugf("Received signal %v", x)
+				header := metadata.Pairs("x-envoy-immediate-health-check-fail", "")
+				grpc.SendHeader(ctx, header)
```

Comment:
we only want to send the health check fail on some signals, i.e. these ones: https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html